spring-webflux/src/main/java/org/springframework/web/reactive/function/BodyInserters.java

```diff
@@ -72,19 +75,86 @@
 	}
 
 	/**
-	 * Inserter to write the given object.
-	 * <p>Alternatively, consider using the {@code syncBody(Object)} shortcuts on
+	 * Inserter to write the given object. The body can be one of the following:
+	 * <ul>
+	 * <li>Concrete value
+	 * <li>{@link Publisher} of value(s)
+	 * <li>Any other producer of value(s) that can be adapted to a
+	 * {@link Publisher} via {@link ReactiveAdapterRegistry}
+	 * </ul>
+	 * <p>Alternatively, consider using the {@code body(Object)} shortcuts on
 	 * {@link org.springframework.web.reactive.function.client.WebClient WebClient} and
 	 * {@link org.springframework.web.reactive.function.server.ServerResponse ServerResponse}.
-	 * @param body the body to write to the response
+	 * @param body the body to write
 	 * @param <T> the type of the body
 	 * @return the inserter to write a single object
 	 */
 	public static <T> BodyInserter<T, ReactiveHttpOutputMessage> fromObject(T body) {
+		Assert.notNull(body, "'body' must not be null");
+		ReactiveAdapter adapter = ReactiveAdapterRegistry.getSharedInstance().getAdapter(body.getClass());
+		ResolvableType bodyType;
+		if (body instanceof Publisher || adapter != null) {
+			bodyType = ResolvableType.forClass(Object.class);
```

Comment:
So if it is async we use Object? Maybe I'm missing something but this doesn't seem right. In RSocketRequester we wait for the actual value in order to determine the type and then encode.

Reply:
As discussed, I don't think this is possible given current `writer.write(publisher)` API, and also as pointed out by Rossen RSocket and WebFlux are different: Payload oriented (one message at a time) vs HTTP producing a complete stream. So I will provide only the variant with explicit types for now.